src/__private_api.rs

```diff
@@ -52,57 +52,43 @@ impl Log for GlobalLogger {
     }
 }
 
-// Split from `log` to reduce generics and code size
-fn log_impl<L: Log>(
-    logger: L,
-    args: Arguments,
-    level: Level,
-    &(target, module_path, loc): &(&str, &'static str, &'static Location),
-    kvs: Option<&[(&str, Value)]>,
-) {
-    #[cfg(not(feature = "kv"))]
-    if kvs.is_some() {
-        panic!("key-value support is experimental and must be enabled using the `kv` feature")
+pub trait LogExt: Log {
+    fn do_enabled(&self, level: Level, target: &str) -> bool {
+        self.enabled(&Metadata::builder().level(level).target(target).build())
     }
 
-    let mut builder = Record::builder();
-
-    builder
-        .args(args)
-        .level(level)
-        .target(target)
-        .module_path_static(Some(module_path))
-        .file_static(Some(loc.file()))
-        .line(Some(loc.line()));
-
-    #[cfg(feature = "kv")]
-    builder.key_values(&kvs);
-
-    logger.log(&builder.build());
-}
-
-pub fn log<'a, K, L>(
-    logger: L,
-    args: Arguments,
-    level: Level,
-    target_module_path_and_loc: &(&str, &'static str, &'static Location),
-    kvs: K,
-) where
-    K: KVs<'a>,
-    L: Log,
-{
-    log_impl(
-        logger,
-        args,
-        level,
-        target_module_path_and_loc,
-        kvs.into_kvs(),
-    )
+    fn do_log<'a>(
```

Comment:
If we’re going to do it this way these names should probably be much more mangled, like `__log_private_api_log`. Similarly, the name of the trait should also be ugly so it doesn’t look like something you’re meant to use directly.

Reply:
Agree. I'd first check whether this way work (so that all the zero-size logger can benefit). And rename later.